util/website.py

```diff
@@ -0,0 +1,189 @@
+                for dir in dirs:
+                    if not sftp.isdir(dir):
+                        print "Creating directory", dir
+                        sftp.makedirs(dir)
+                        created.append(dir)
+
+                # Push all files, chdir() for each
+                for fil in files:
+                    with sftp.cd(os.path.dirname(fil)):
+                        print "Pushing", fil
+                        sftp.put(fil)
+                        pushed.append(fil)
+
+        return pushed
+    except paramiko.ssh_exception.AuthenticationException as ex:
+        print "Authentication error, please verify credentials"
+        return False
+    except IOError as ex:
+        print "Unable to read a file (private key or invalid path on server ?)", ex
+        return False
+
+    # Should not be reached
+    return False
+
+def maybe_publish(file='index.htm'):
+    """
+    Publishing to the web requires the following env variables to be set
+        'ds_website_username: defines the SSH username to connect to
+        'ds_website_privkey: defines the SSH privkey filename to use for connection
+        'ds_website_server_fqdn: hostname of the SSH server
+        'ds_website_server_port: port of the SSH server, defaults to 22
+        'ds_website_server_root: directory on the server where to push data,
+                               should be either empty, or containing at least
+                               a .htaccess and index.htm file
+    """
+
+    ssh_auth_infos = {
+        'ds_website_username': '',
+        'ds_website_privkey': '',
+        'ds_website_server_fqdn': '',
+        'ds_website_server_port': 22,
+        'ds_website_server_root': ''
+    }
+
+    for key in ssh_auth_infos.keys():
+       vartype = type(ssh_auth_infos[key])
+       value = os.environ.get(key)
+       if value is not None:
+           if vartype == str:
+               ssh_auth_infos[key] = str(os.environ.get(key))
+           elif vartype == int:
+               try:
+                   ssh_auth_infos[key] = int(os.environ.get(key))
+               except TypeError as ex:
+                   print "WARNING:", "Keeping default SSH port value because of:", ex
+
+    missing_env = filter(lambda x: len(str(ssh_auth_infos[x])) == 0, ssh_auth_infos.keys())
+    if len(missing_env) > 0:
+        print "Not publishing, missing some required environment variables:", missing_env
+        print "But maybe this is what you wanted, after all ..."
+        return False
+
+    merge_logs("logs")
+
+    all_deps = parse_for_deps(file)
+
+    if len(all_deps) == 0:
+        print "Problem during deps computation, aborting"
+        return False
+
+    if not verify_ssh_dir(ssh_auth_infos):
+        print "Problem during SSH directory verification, aborting"
+        return False
+
+    all_files = [ file ] + all_deps
+    uploaded = push_files_sftp(all_files, ssh_auth_infos)
```

Comment:
Is this supposed to be reentrant? If you ran ```maybe_publish``` for the first time on an empty web root dir, the web root dir will get populated. Calling it the second time could fail due to ```verify_ssh_dir``` not finding an empty directory or one with ```index.htm``` and ```.htaccess``` in it. At least, if you are not assuming ```index.htm``` and ```.htaccess``` to always be part of the transferred data. Which kind of problem do you want to avoid by this approach/```verify_ssh_dir```?

Do we ever get rid of outdated files (former dependencies) in the web root? 
How about wiping it before transfer?